marshal/unmarshal.go

```diff
@@ -0,0 +1,122 @@
+package marshal
+
+import (
+	"encoding/json"
+	"errors"
+	"reflect"
+)
+
+// Unmarshals a JSON element and fills in the ForceSendFields field if
+// the struct contains it. Only anotates basic types in the ForceSendFields.
+func Unmarshal(data []byte, v any) error {
+	if len(data) == 0 {
+		return nil
+	}
+	// create a new element if the pointer is nul
+	if v == nil {
+		objectType := reflect.ValueOf(v).Type()
+		v = reflect.New(objectType).Interface()
+	}
+	err := unmarshalInternal(data, v)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setForceSendFields(v any, presentFields []string) error {
+
+	value := reflect.ValueOf(v)
+	if value.Kind() != reflect.Ptr {
+		return nil
+	}
+
+	// reflex.GetFieldByName may return the field from an anonymous field
+	// if the extending struct doesn't have the field.
+	field := getFieldByName(v, forceSendFieldName)
+	if !field.IsValid() {
+		return nil
+	}
+
+	if !field.CanSet() || field.Kind() != reflect.Slice {
+		return errors.New("cannot set field")
```

Comment:
When can this happen?

Reply:
It should not happen, or happen by mistake:
* [Can't happen] The field is private (then it won't match the name)
* [Shold not happen] The field is not addressable
* [Mistake] Field has the wrong type